staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/cbor.go

```diff
@@ -0,0 +1,304 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cbor
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"reflect"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer/recognizer"
+	util "k8s.io/apimachinery/pkg/util/runtime"
+
+	"github.com/fxamacker/cbor/v2"
+)
+
+var encMode cbor.EncMode = func() cbor.EncMode {
+	encMode, err := cbor.EncOptions{
+		// Map keys need to be sorted to have deterministic output, and this is the order
+		// defined in RFC 8949 4.2.1 "Core Deterministic Encoding Requirements".
+		Sort: cbor.SortBytewiseLexical,
+
+		// CBOR supports distinct types for IEEE-754 float16, float32, and float64. Store
+		// floats in the smallest width that preserves value so that equivalent float32 and
+		// float64 values encode to identical bytes, as they do in a JSON
+		// encoding. Satisfies one of the "Core Deterministic Encoding Requirements".
+		ShortestFloat: cbor.ShortestFloat16,
+		NaNConvert:    cbor.NaNConvert7e00,
+		InfConvert:    cbor.InfConvertFloat16,
+
+		// Prefer encoding math/big.Int to one of the 64-bit integer types if it fits.
+		BigIntConvert: cbor.BigIntConvertShortest,
+
+		// MarshalJSON for time.Time writes RFC3339 with nanos.
+		Time: cbor.TimeRFC3339Nano,
+
+		// Don't add tag 1 when encoding time.Time. For JSON parity, we can't require it
+		// when decoding to time.Time (in manifests and in dynamic client code it looks like
+		// any string), and decoding to interface{} must produce a string.
+		TimeTag: cbor.EncTagNone,
```

Comment:
somewhat surprised by this.  Is there a specific test indicating the failure mode on this?